packages/jupyter-ai/jupyter_ai/tools/models.py

```diff
@@ -0,0 +1,213 @@
+        >>> tool.read
+        True
+    """
+
+    callable: Callable = Field(exclude=True)
+    name: Optional[str] = None
+    description: Optional[str] = None
+    read: bool = False
+    write: bool = False
+    execute: bool = False
+    delete: bool = False
+
+    @model_validator(mode="after")
+    def set_name_description(self):
+        if not self.name:
+            if hasattr(self.callable, "__name__") and self.callable.__name__:
+                self.name = self.callable.__name__
+            else:
+                raise ValueError("Unable to extract name from callable")
+
+        if not self.description:
+            self.description = get_doc_description(self.callable)
+
+        return self
+
+    def __eq__(self, other):
+        if not isinstance(other, Tool):
+            return False
+        return (
+            self.name == other.name
+            and self.description == other.description
+            and self.read == other.read
+            and self.write == other.write
+            and self.delete == other.delete
+            and self.execute == other.execute
+        )
+
+    def __hash__(self):
+        return hash(
+            (
+                self.name,
+                self.description,
+                self.read,
+                self.write,
+                self.delete,
+                self.execute,
+            )
+        )
+
+
+class Toolkit(BaseModel):
+    """A collection of tools with capability-based filtering.
+
+    A Toolkit groups related tools together and provides methods to add tools
+    and find tools based on their capability flags.
+
+    Attributes:
+        name: The toolkit's name.
+        description: Optional description of the toolkit.
+        tools: The set of tools in this toolkit.
+
+    Example:
+        >>> def read_file(path: str):
+        ...     '''Read a file from disk.'''
+        ...     with open(path) as f:
+        ...         return f.read()
+        >>> def write_file(path: str, content: str):
+        ...     '''Write content to a file.'''
+        ...     with open(path, 'w') as f:
+        ...         f.write(content)
+        >>> toolkit = Toolkit(name="FileTools", description="File operations")
+        >>> read_tool = Tool(callable=read_file, read=True)
+        >>> write_tool = Tool(callable=write_file, write=True)
+        >>> toolkit.add_tool(read_tool)
+        >>> toolkit.add_tool(write_tool)
+        >>> len(toolkit.tools)
+        2
+        >>> read_tools = toolkit.find_tools(read=True)
+        >>> len(read_tools)
+        1
+    """
+
+    name: str
+    description: Optional[str] = None
+    tools: set = Field(default_factory=set)
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def add_tool(self, tool: Tool):
+        """Add a tool to this toolkit.
+
+        Args:
+            tool: The tool to add to the toolkit.
+        """
+        self.tools.add(tool)
+
+    def get_tools(
+        self,
+        read: bool = False,
```

Comment:
Let's replace the default values here by `None` which will indicate that either True or False for that attribute are fine.